qualtran/bloqs/mcmt/multi_control_multi_target_pauli.py

```diff
@@ -111,43 +101,61 @@ class MultiControlPauli(GateWithRegisters):
     target_gate: cirq.Pauli = cirq.X
 
     @cached_property
-    def signature(self) -> Signature:
-        return Signature.build(controls=len(self.cvs), target=1)
+    def signature(self) -> 'Signature':
+        ctrl = Register('controls', QBit(), shape=(len(self.cvs),))
+        target = Register('target', QBit())
+        return Signature([ctrl, target] if len(self.cvs) > 0 else [target])
 
     def decompose_from_registers(
         self, *, context: cirq.DecompositionContext, **quregs: NDArray['cirq.Qid']
     ) -> cirq.OP_TREE:
-        controls, target = quregs.get('controls', ()), quregs['target']
-        if len(self.cvs) < 2:
+        controls, target = quregs.get('controls', np.array([])), quregs['target']
+        if len(self.cvs) <= 2:
+            controls = controls.flatten()
+            yield [cirq.X(q) for cv, q in zip(self.cvs, controls) if cv == 0]
             yield self.target_gate.on(*target).controlled_by(*controls)
+            yield [cirq.X(q) for cv, q in zip(self.cvs, controls) if cv == 0]
             return
         qm = context.qubit_manager
         and_ancilla, and_target = np.array(qm.qalloc(len(self.cvs) - 2)), qm.qalloc(1)
-        ctrl, junk = controls[:, np.newaxis], and_ancilla[:, np.newaxis]
-        if len(self.cvs) == 2:
-            and_op = And(*self.cvs).on_registers(ctrl=ctrl, target=and_target)
-        else:
-            and_op = MultiAnd(self.cvs).on_registers(ctrl=ctrl, junk=junk, target=and_target)
+        and_op = MultiAnd(self.cvs).on_registers(
+            ctrl=controls, junk=and_ancilla[:, np.newaxis], target=and_target
+        )
         yield and_op
         yield self.target_gate.on(*target).controlled_by(*and_target)
         yield and_op**-1
         qm.qfree([*and_ancilla, *and_target])
 
     def short_name(self) -> str:
-        return r'$C^{n}(P)$'
+        n = len(self.cvs)
+        ctrl = f'C^{n}' if n > 2 else ['', 'C', 'CC'][n]
+        return f'{ctrl}{self.target_gate!s}'
 
     def _circuit_diagram_info_(self, _) -> cirq.CircuitDiagramInfo:
         wire_symbols = ["@" if b else "@(0)" for b in self.cvs]
         wire_symbols += [str(self.target_gate)]
         return cirq.CircuitDiagramInfo(wire_symbols=wire_symbols)
 
+    def on_classical_vals(
+        self, controls: 'ClassicalValT', target: 'ClassicalValT'
```

Comment:
this assumes >= 1 controls

Reply:
Added support for 0 controls and a corresponding test